stdlib/Profile/src/heapsnapshot_reassemble.jl

```diff
@@ -0,0 +1,163 @@
+            buf[i] = (d100 >> 0x8) % UInt8
+            x = oftype(x, d)
+            i -= 2
+        end
+        if i > 0
+            @inbounds buf[i] = 0x30 + (rem(x, 0xa) % UInt8)::UInt8
+        end
+        write(io, @view buf[max(i, 1):n])
+    end
+end
+
+function assemble_snapshot(in_prefix, out_file::AbstractString)
+    open(out_file, "w") do io
+        assemble_snapshot(in_prefix, io)
+    end
+end
+# Manually parse and write the .json files, given that we don't have JSON import/export in
+# julia's stdlibs.
+function assemble_snapshot(in_prefix, io::IO)
+    preamble = read(string(in_prefix, ".json"), String)
+    pos = last(findfirst("node_count\":", preamble)) + 1
+    endpos = findnext(==(','), preamble, pos) - 1
+    node_count = parse(Int, String(@view preamble[pos:endpos]))
+
+    pos = last(findnext("edge_count\":", preamble, endpos)) + 1
+    endpos = findnext(==('}'), preamble, pos) - 1
+    edge_count = parse(Int, String(@view preamble[pos:endpos]))
+
+    nodes = init_nodes(node_count, edge_count)
+
+    # Parse nodes with empty edge counts that we need to fill later
+    # TODO: preallocate line buffer
+    for (i, line) in enumerate(eachline(string(in_prefix, ".nodes")))
+        iter = eachsplit(line, ',')
+        x, s = iterate(iter)
+        node_type = parse(Int8, x)
+        x, s = iterate(iter, s)
+        node_name_index = parse(UInt32, x)
+        x, s = iterate(iter, s)
+        id = parse(UInt, x)
+        x, s = iterate(iter, s)
+        self_size = parse(Int,  x)
+        x, s = iterate(iter, s)
+        edge_count = parse(UInt, x)
+        @assert edge_count == 0
+        x, s = iterate(iter, s)
+        @assert parse(Int8, x) == 0 # trace_node_id
+        x, s = iterate(iter, s)
+        @assert parse(Int8, x) == 0 # detachedness
+
+        nodes.type[i] = node_type
+        nodes.name_index[i] = node_name_index
+        nodes.id[i] = id
+        nodes.self_size[i] = self_size
+        nodes.edge_count[i] = edge_count
+    end
+
+    # Parse the edges to fill in the edge counts for nodes and correct the to_node offsets
+    # TODO: preallocate line buffer
+    for (i, line) in enumerate(eachline(string(in_prefix, ".edges")))
+        iter = eachsplit(line, ',')
+        x, s = iterate(iter)
+        edge_type = parse(Int8, x)
+        x, s = iterate(iter, s)
+        edge_name_index = parse(UInt, x)
+        x, s = iterate(iter, s)
+        to_node = parse(UInt32, x)
+        x, s = iterate(iter, s)
+        from_node = parse(Int,  x)
+
+        nodes.edges.type[i] = edge_type
+        nodes.edges.name_index[i] = edge_name_index
+        nodes.edges.to_pos[i] = to_node * 7 # 7 fields per node, the streaming format doesn't multiply the offset by 7
+        nodes.edge_count[from_node] += UInt32(1)
+    end
+
+    _digits_buf = zeros(UInt8, ndigits(typemax(UInt)))
+    println(io, @view(preamble[1:end-2]), ",") # remove trailing "}\n", we don't end the snapshot here
+    println(io, "\"nodes\":[")
```

Comment:
The nodes and edges files currently aren't valid json, since we're not writing out the leading and trailing `[]`, and we aren't writing a trailing comma after each line.

I figured it was easier to process this way, which seems to bear out with the code you wrote, @drvi. But now I'm kind of thinking we may as well have every file we output be valid JSON...?

On the other hand, that's an extra character per node and per edge, of which there can be _billions_, so this could add a whole GiB right there?

Reply:
If this is adding a GB, then the whole file is probably 10s of GB, since that one character will only be a small fraction of each line. But maybe we want to consider output these as BSON instead, to get faster encoding performance for them.

Reply:
In retrospect, i think these are actually truly csvs! I'm going to rename the output files to be `.csv`. I think that at least makes them self-documenting in their format.

----

Regarding the streaming time and file size... I agree, writing them out as binary files would be even better! 🤔 They're literally just 2 giant matrices of numbers like:
```
$ cat 58178_136122584059625.heapsnapshot.edges  | head -n 3
0,2,0,1
0,3,0,2
0,2,0,3

$ cat 58178_136122584059625.heapsnapshot.nodes  | head -n 3
0,0,0,0,0,0,0
1,1,4482646032,384,0,0,0
1,1,15281513072,384,0,0,0
```
So we should definitely consider some kind of binary format instead. 👍

Reply:
Is there something even simpler than BSON for this? Like, i think we could literally just output an array of int64 binary data, and then read it in like that, no?

Reply:
Heh, so it turns out that writing these as binary data _does_ make it faster, but the files are actually slightly *bigger*, since i guess a lot of the indexes were smallish numbers (so only a few bytes) whereas their binary format is the full 8 bytes.

But still, the speed probably makes this worth it 👍

Reply:
A couple of thoughts on the binary format. Since the schema is known, it wouldn't be hard to handle each column separately in the snapshot assembler.

For `nodes` we know that `type` is an index into a small array of `node_types` so that could be a single byte (or less), `name` and `self_size` won't need 8 bytes either, and would benefit from varint (aka vbyte) encoding like protobuf does (this eliminates the leading zeros and is relatively easy to implement). `edge_count`, `trace_node_id` and `detachedness` are always zero so they can be omitted altogether. `id` is an interesting case as currently it's the pointer to the object, but once written to a file, we can just enumerate the nodes to get a unique id so we don't need to write the pointer out (but maybe having all the pointers in your program could be useful for some analysis?).

For `edges`, again `type` could be a single byte (or less). `name_or_index` (the index into `strings` for some edge types) is interesting as we could use varint encoding _but_  I think we are using `typemax(UInt64)` for edge types that don't have a corresponding `string` which wouldn't compress in varint (it would get bigger in fact) so maybe for these edge types we shouldn't encode any `name_or_index` value at all or have a special value for those. `to_node` and `from_node` are bounded by the size of `nodes` so varint should help again.

One issue with the current approach is that in order to reassemble the snapshot, one needs to update the `edge_count` for each node, which means that you need to have all `nodes` in memory and you need to iterate the `edges` twice (once to update nodes, and then to write them to the assembled snapshot). How about we produce another file, `edge_counts`, which would basically be an array of edge counts for each node that we accumulate and write at the end (for 10M nodes that would be 40MB)?

Reply:
You don't just need the edge_counts; the edges need to be grouped and ordered by the nodes. So you read the file by iterating the nodes, seeing how many edges it has, and then those first N edges are coming out of the first node, then you move to the second node, and the next M edges are coming from that node, etc.

I dunno if you can build that file without having all the nodes in memory while we iterate the edges? Maybe I'm not following what you were explaining?

Reply:
Ah my bad, I thought the edges were written out already ordered, yeah then the `edge_count` idea doesn't apply.

Reply:
Mmm yeah makes sense. Yeah it's too bad. :(

Accounting for that was the last bug that I fixed while you were out on vacation. The file format is really gnarly and i think we didn't get to this detail when we talked through it way back at the start. 😊